[processor/routing] Collector crash

### Component(s)

processor/routing

### What happened?

## Description

Collector crash, caused by misuse of the ptrace.Span construct by the routing processor.

## Steps to Reproduce

1) start the collector with the provided config
2) run `telemetrygen traces --traces 10000 --otlp-insecure --otlp-attributes='attr1="demo",attr2="xxx"'`
3) watch it burn

## Expected Result

No crashes.

## Actual Result

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb1d720]

goroutine 196 [running]:
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
        go.opentelemetry.io/otel/sdk@v1.16.0/trace/span.go:383 +0x25
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc002e7a180, {0x0, 0x0, 0xd5c01e0?})
        go.opentelemetry.io/otel/sdk@v1.16.0/trace/span.go:415 +0x975
panic({0x698d520?, 0xd52b350?})
        runtime/panic.go:914 +0x21f
go.opentelemetry.io/collector/pdata/ptrace.Span.Attributes({0x0?})
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/ptrace/generated_span.go:124
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal.SpanPathGetSetter[...].func31({{0x0?}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/contexts/internal/span.go:408 +0x5b
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.StandardGetSetter[...].Get(...)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/expression.go:49
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Parser[...]).newComparisonEvaluator.func2({{0x0}, {0x0}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:87 +0xcd
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.BoolExpr[...].Eval(0xc001114fa0?, {0x88e5f18?, 0xc001d90ea0?}, {{0x0?}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:19 +0x47
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Parser[...].func2({{0x0}, {0x0}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:43 +0xec
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.BoolExpr[...].Eval(0x1?, {0x88e5f18?, 0xc001d90ea0?}, {{0x0?}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:19 +0x47
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Parser[...].func2({{0x0}, {0x0}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:60 +0xec
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.BoolExpr[...].Eval(...)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:19
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Statement[...]).Execute(0xc002e87120, {0x88e5f18?, 0xc001d90ea0}, {{0x0}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/parser.go:55 +0x76
github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor.(*tracesProcessor).route(0xc002932460, {0x88e5f18, 0xc001d90ea0}, {0x7a7910f?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor@v0.84.0/traces.go:116 +0x3db
github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor.(*tracesProcessor).ConsumeTraces(0x0?, {0x88e5f18?, 0xc001d90ea0?}, {0x6992e00?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor@v0.84.0/traces.go:81 +0x28
go.opentelemetry.io/collector/consumer.ConsumeTracesFunc.ConsumeTraces(...)
        go.opentelemetry.io/collector/consumer@v0.84.0/traces.go:25
go.opentelemetry.io/collector/receiver/otlpreceiver/internal/trace.(*Receiver).Export(0xc002925470, {0x88e5f18, 0xc001d90e10}, {0x0?})
        go.opentelemetry.io/collector/receiver/otlpreceiver@v0.84.0/internal/trace/otlp.go:41 +0xc5
go.opentelemetry.io/collector/pdata/ptrace/ptraceotlp.rawTracesServer.Export({{0x889a548?, 0xc002925470?}}, {0x88e5f18?, 0xc001d90e10?}, 0xc002e7a100?)
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/ptrace/ptraceotlp/grpc.go:83 +0xb0
go.opentelemetry.io/collector/pdata/internal/data/protogen/collector/trace/v1._TraceService_Export_Handler.func1({0x88e5f18, 0xc001d90e10}, {0x769a460?, 0xc002b00a08})
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/internal/data/protogen/collector/trace/v1/trace_service.pb.go:310 +0x72
go.opentelemetry.io/collector/config/configgrpc.(*GRPCServerSettings).toServerOption.enhanceWithClientInformation.func9({0x88e5f18?, 0xc001d90db0?}, {0x769a460, 0xc002b00a08}, 0xa?, 0xc002b00cd8)
        go.opentelemetry.io/collector/config/configgrpc@v0.84.0/configgrpc.go:400 +0x46
google.golang.org/grpc.getChainUnaryHandler.func1({0x88e5f18, 0xc001d90db0}, {0x769a460, 0xc002b00a08})
        google.golang.org/grpc@v1.57.0/server.go:1179 +0xb2
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1({0x88e5f18, 0xc001d907e0}, {0x769a460, 0xc002b00a08}, 0xc002e692a0, 0xc002e73100)
        go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.42.1-0.20230612162650-64be7e574a17/interceptor.go:344 +0x52b
google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0x88e5f18, 0xc001d907e0}, {0x769a460, 0xc002b00a08}, 0xc002088a50?, 0x698d9a0?)
        google.golang.org/grpc@v1.57.0/server.go:1170 +0x85
go.opentelemetry.io/collector/pdata/internal/data/protogen/collector/trace/v1._TraceService_Export_Handler({0x683fd60?, 0xc00293d060}, {0x88e5f18, 0xc001d907e0}, 0xc001faf180, 0xc00292d4a0)
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/internal/data/protogen/collector/trace/v1/trace_service.pb.go:312 +0x135
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00293a1e0, {0x8916aa0, 0xc0027009c0}, 0xc002923320, 0xc002966d50, 0xd539870, 0x0)
        google.golang.org/grpc@v1.57.0/server.go:1360 +0xe15
google.golang.org/grpc.(*Server).handleStream(0xc00293a1e0, {0x8916aa0, 0xc0027009c0}, 0xc002923320, 0x0)
        google.golang.org/grpc@v1.57.0/server.go:1737 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc@v1.57.0/server.go:982 +0x8d
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 195
        google.golang.org/grpc@v1.57.0/server.go:980 +0x165
```

### Collector version

0.84.0

### Environment information

## Environment
OS: Fedora release 38 (Thirty Eight)
 

### OpenTelemetry Collector configuration

```yaml
extensions:

receivers:
  otlp:
    protocols:
      grpc:

processors:
  routing:
    default_exporters: [logging]
    error_mode: ignore
    table:
      - statement: route() where attributes["attr1"] == "demo"
        exporters: [ logging/route-1 ]
      - statement: route() where attributes["attr1"] == "demo" and attributes["attr2"] == "xxx"
        exporters: [ logging/route-2 ]

exporters:
  logging:
  logging/route-1:
  logging/route-2:

connectors:

service:
  extensions:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [routing]
      exporters: [logging, logging/route-1, logging/route-2]
```


### Log output

```shell
2023-09-05T07:02:35.045-0300    info    service/telemetry.go:84 Setting up own telemetry...
2023-09-05T07:02:35.045-0300    info    service/telemetry.go:201        Serving Prometheus metrics      {"address": ":8888", "level": "Basic"}
2023-09-05T07:02:35.045-0300    info    exporter@v0.84.0/exporter.go:275        Development component. May change in the future.        {"kind": "exporter", "data_type": "traces", "name": "logging/route-2"}
2023-09-05T07:02:35.045-0300    info    exporter@v0.84.0/exporter.go:275        Development component. May change in the future.        {"kind": "exporter", "data_type": "traces", "name": "logging"}
2023-09-05T07:02:35.045-0300    info    exporter@v0.84.0/exporter.go:275        Development component. May change in the future.        {"kind": "exporter", "data_type": "traces", "name": "logging/route-1"}
2023-09-05T07:02:35.047-0300    info    service/service.go:138  Starting otelcol-contrib...     {"Version": "0.84.0", "NumCPU": 16}
2023-09-05T07:02:35.047-0300    info    extensions/extensions.go:31     Starting extensions...
2023-09-05T07:02:35.048-0300    warn    internal@v0.84.0/warning.go:40  Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks        {"kind": "receiver", "name": "otlp", "data_type": "traces", "documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
2023-09-05T07:02:35.048-0300    info    otlpreceiver@v0.84.0/otlp.go:83 Starting GRPC server    {"kind": "receiver", "name": "otlp", "data_type": "traces", "endpoint": "0.0.0.0:4317"}
2023-09-05T07:02:35.048-0300    info    service/service.go:161  Everything is ready. Begin running and processing data.
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb1d720]

goroutine 196 [running]:
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
        go.opentelemetry.io/otel/sdk@v1.16.0/trace/span.go:383 +0x25
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc002e7a180, {0x0, 0x0, 0xd5c01e0?})
        go.opentelemetry.io/otel/sdk@v1.16.0/trace/span.go:415 +0x975
panic({0x698d520?, 0xd52b350?})
        runtime/panic.go:914 +0x21f
go.opentelemetry.io/collector/pdata/ptrace.Span.Attributes({0x0?})
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/ptrace/generated_span.go:124
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal.SpanPathGetSetter[...].func31({{0x0?}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/contexts/internal/span.go:408 +0x5b
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.StandardGetSetter[...].Get(...)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/expression.go:49
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Parser[...]).newComparisonEvaluator.func2({{0x0}, {0x0}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:87 +0xcd
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.BoolExpr[...].Eval(0xc001114fa0?, {0x88e5f18?, 0xc001d90ea0?}, {{0x0?}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:19 +0x47
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Parser[...].func2({{0x0}, {0x0}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:43 +0xec
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.BoolExpr[...].Eval(0x1?, {0x88e5f18?, 0xc001d90ea0?}, {{0x0?}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:19 +0x47
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Parser[...].func2({{0x0}, {0x0}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:60 +0xec
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.BoolExpr[...].Eval(...)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/boolean_value.go:19
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl.(*Statement[...]).Execute(0xc002e87120, {0x88e5f18?, 0xc001d90ea0}, {{0x0}, {0x0?}, {0xc002991278?}, {0xc002b00d20?}})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl@v0.84.0/parser.go:55 +0x76
github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor.(*tracesProcessor).route(0xc002932460, {0x88e5f18, 0xc001d90ea0}, {0x7a7910f?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor@v0.84.0/traces.go:116 +0x3db
github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor.(*tracesProcessor).ConsumeTraces(0x0?, {0x88e5f18?, 0xc001d90ea0?}, {0x6992e00?})
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor@v0.84.0/traces.go:81 +0x28
go.opentelemetry.io/collector/consumer.ConsumeTracesFunc.ConsumeTraces(...)
        go.opentelemetry.io/collector/consumer@v0.84.0/traces.go:25
go.opentelemetry.io/collector/receiver/otlpreceiver/internal/trace.(*Receiver).Export(0xc002925470, {0x88e5f18, 0xc001d90e10}, {0x0?})
        go.opentelemetry.io/collector/receiver/otlpreceiver@v0.84.0/internal/trace/otlp.go:41 +0xc5
go.opentelemetry.io/collector/pdata/ptrace/ptraceotlp.rawTracesServer.Export({{0x889a548?, 0xc002925470?}}, {0x88e5f18?, 0xc001d90e10?}, 0xc002e7a100?)
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/ptrace/ptraceotlp/grpc.go:83 +0xb0
go.opentelemetry.io/collector/pdata/internal/data/protogen/collector/trace/v1._TraceService_Export_Handler.func1({0x88e5f18, 0xc001d90e10}, {0x769a460?, 0xc002b00a08})
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/internal/data/protogen/collector/trace/v1/trace_service.pb.go:310 +0x72
go.opentelemetry.io/collector/config/configgrpc.(*GRPCServerSettings).toServerOption.enhanceWithClientInformation.func9({0x88e5f18?, 0xc001d90db0?}, {0x769a460, 0xc002b00a08}, 0xa?, 0xc002b00cd8)
        go.opentelemetry.io/collector/config/configgrpc@v0.84.0/configgrpc.go:400 +0x46
google.golang.org/grpc.getChainUnaryHandler.func1({0x88e5f18, 0xc001d90db0}, {0x769a460, 0xc002b00a08})
        google.golang.org/grpc@v1.57.0/server.go:1179 +0xb2
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1({0x88e5f18, 0xc001d907e0}, {0x769a460, 0xc002b00a08}, 0xc002e692a0, 0xc002e73100)
        go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.42.1-0.20230612162650-64be7e574a17/interceptor.go:344 +0x52b
google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0x88e5f18, 0xc001d907e0}, {0x769a460, 0xc002b00a08}, 0xc002088a50?, 0x698d9a0?)
        google.golang.org/grpc@v1.57.0/server.go:1170 +0x85
go.opentelemetry.io/collector/pdata/internal/data/protogen/collector/trace/v1._TraceService_Export_Handler({0x683fd60?, 0xc00293d060}, {0x88e5f18, 0xc001d907e0}, 0xc001faf180, 0xc00292d4a0)
        go.opentelemetry.io/collector/pdata@v1.0.0-rcv0014/internal/data/protogen/collector/trace/v1/trace_service.pb.go:312 +0x135
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00293a1e0, {0x8916aa0, 0xc0027009c0}, 0xc002923320, 0xc002966d50, 0xd539870, 0x0)
        google.golang.org/grpc@v1.57.0/server.go:1360 +0xe15
google.golang.org/grpc.(*Server).handleStream(0xc00293a1e0, {0x8916aa0, 0xc0027009c0}, 0xc002923320, 0x0)
        google.golang.org/grpc@v1.57.0/server.go:1737 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc@v1.57.0/server.go:982 +0x8d
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 195
        google.golang.org/grpc@v1.57.0/server.go:980 +0x165
```


### Additional context

Logs from telemetrygen:

```
telemetrygen traces --traces 10000 --otlp-insecure --otlp-attributes='attr1="demo",attr2="xxx"'
2023-09-05T07:02:46.801-0300    INFO    traces/traces.go:62     starting gRPC exporter
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/clientconn.go:477  [core][Channel #1] Channel created      {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/clientconn.go:1786 [core][Channel #1] original dial target is: "localhost:4317"    {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/clientconn.go:1793 [core][Channel #1] parsed dial target is: {URL:{Scheme:localhost Opaque:4317 User: Host: Path: RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}      {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/clientconn.go:1807 [core][Channel #1] fallback to scheme "passthrough"     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/clientconn.go:1815 [core][Channel #1] parsed dial target is: {URL:{Scheme:passthrough Opaque: User: Host: Path:/localhost:4317 RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}} {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/clientconn.go:1948 [core][Channel #1] Channel authority set to "localhost:4317"    {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.801-0300    INFO    grpc@v1.57.0/resolver_conn_wrapper.go:238       [core][Channel #1] Resolver state updated: {
  "Addresses": [
    {
      "Addr": "localhost:4317",
      "ServerName": "",
      "Attributes": null,
      "BalancerAttributes": null,
      "Type": 0,
      "Metadata": null
    }
  ],
  "ServiceConfig": null,
  "Attributes": null
} (resolver returned new addresses)     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.802-0300    INFO    grpc@v1.57.0/balancer_conn_wrappers.go:192      [core][Channel #1] Channel switches to new LB policy "pick_first"       {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.802-0300    INFO    grpc@v1.57.0/balancer_conn_wrappers.go:312      [core][Channel #1 SubChannel #2] Subchannel created     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.802-0300    INFO    grpc@v1.57.0/clientconn.go:565  [core][Channel #1] Channel Connectivity change to CONNECTING    {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.802-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.802-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.804-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to READY        {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.804-0300    INFO    grpc@v1.57.0/clientconn.go:565  [core][Channel #1] Channel Connectivity change to READY {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.804-0300    INFO    traces/traces.go:120    generation of traces isn't being throttled
2023-09-05T07:02:46.828-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.828-0300    INFO    grpc@v1.57.0/clientconn.go:565  [core][Channel #1] Channel Connectivity change to IDLE  {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:46.899-0300    INFO    traces/worker.go:92     traces generated        {"worker": 0, "traces": 10000}
2023-09-05T07:02:46.899-0300    INFO    traces/traces.go:80     stop the batch span processor
2023-09-05T07:02:53.608-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:53.608-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:53.608-0300    INFO    grpc@v1.57.0/clientconn.go:565  [core][Channel #1] Channel Connectivity change to CONNECTING    {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:53.609-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:02:53.609-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:53.609-0300    INFO    grpc@v1.57.0/clientconn.go:565  [core][Channel #1] Channel Connectivity change to TRANSIENT_FAILURE     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:54.610-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:54.610-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:54.610-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:54.612-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:02:54.612-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:56.319-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:56.319-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:56.319-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:56.321-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:02:56.321-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023/09/05 07:02:56 traces export: context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"
2023-09-05T07:02:59.003-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:59.003-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:59.003-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:02:59.005-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:02:59.005-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:03.742-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:03.742-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:03.742-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:03.743-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:03:03.744-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023/09/05 07:03:06 traces export: context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"
2023-09-05T07:03:10.890-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:10.890-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:10.890-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:10.892-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:03:10.892-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023/09/05 07:03:16 traces export: context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"
2023-09-05T07:03:21.524-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to IDLE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:21.524-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:21.524-0300    INFO    grpc@v1.57.0/clientconn.go:1414 [core][Channel #1 SubChannel #2] Subchannel picks a new address "localhost:4317" to connect     {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:21.525-0300    WARN    grpc@v1.57.0/clientconn.go:1476 [core][Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "localhost:4317", ServerName: "localhost:4317", }. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"    {"system": "grpc", "grpc_log": true}
google.golang.org/grpc.(*addrConn).createTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1476
google.golang.org/grpc.(*addrConn).tryAllAddrs
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1416
google.golang.org/grpc.(*addrConn).resetTransport
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:1353
google.golang.org/grpc.(*addrConn).connect
        /home/jpkroehling/go/pkg/mod/google.golang.org/grpc@v1.57.0/clientconn.go:993
2023-09-05T07:03:21.525-0300    INFO    grpc@v1.57.0/clientconn.go:1303 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"      {"system": "grpc", "grpc_log": true}
2023/09/05 07:03:26 traces export: context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"
2023/09/05 07:03:36 traces export: context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:4317: connect: connection refused"
2023-09-05T07:03:36.824-0300    INFO    grpc@v1.57.0/clientconn.go:565  [core][Channel #1] Channel Connectivity change to SHUTDOWN      {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:36.825-0300    INFO    grpc@v1.57.0/balancer_conn_wrappers.go:203      [core][Channel #1] ccBalancerWrapper: closing   {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:36.825-0300    INFO    grpc@v1.57.0/resolver_conn_wrapper.go:126       [core][Channel #1] Closing the name resolver    {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:36.825-0300    INFO    grpc@v1.57.0/clientconn.go:1301 [core][Channel #1 SubChannel #2] Subchannel Connectivity change to SHUTDOWN     {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:36.825-0300    INFO    grpc@v1.57.0/clientconn.go:1249 [core][Channel #1 SubChannel #2] Subchannel deleted     {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:36.825-0300    INFO    grpc@v1.57.0/clientconn.go:1251 [core][Channel #1] Channel deleted      {"system": "grpc", "grpc_log": true}
2023-09-05T07:03:36.825-0300    INFO    traces/traces.go:70     stopping the exporter
```

--- a/.chloggen/26464-routingprocessor-fix.yaml
+++ b/.chloggen/26464-routingprocessor-fix.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'bug_fix'
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: processor/routing
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: When using attributes instead of resource attributes, the routing processor would crash the collector. This does not affect the connector version of this component.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [26462]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: [user]

--- a/processor/routingprocessor/logs.go
+++ b/processor/routingprocessor/logs.go
@@ -105,8 +105,8 @@ func (p *logProcessor) route(ctx context.Context, l plog.Logs) error {
 	for i := 0; i < l.ResourceLogs().Len(); i++ {
 		rlogs := l.ResourceLogs().At(i)
 		ltx := ottllog.NewTransformContext(
-			plog.LogRecord{},
-			pcommon.InstrumentationScope{},
+			plog.NewLogRecord(),
+			pcommon.NewInstrumentationScope(),
 			rlogs.Resource(),
 		)
 

--- a/processor/routingprocessor/logs_test.go
+++ b/processor/routingprocessor/logs_test.go
@@ -400,6 +400,47 @@ func TestLogsAreCorrectlySplitPerResourceAttributeWithOTTL(t *testing.T) {
 	})
 }
 
+// see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26462
+func TestLogsAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
+	// prepare
+	defaultExp := &mockLogsExporter{}
+	firstExp := &mockLogsExporter{}
+
+	host := newMockHost(map[component.DataType]map[component.ID]component.Component{
+		component.DataTypeLogs: {
+			component.NewID("otlp"):              defaultExp,
+			component.NewIDWithName("otlp", "1"): firstExp,
+		},
+	})
+
+	exp, err := newLogProcessor(noopTelemetrySettings, &Config{
+		DefaultExporters: []string{"otlp"},
+		Table: []RoutingTableItem{
+			{
+				Statement: `route() where attributes["value"] > 0`,
+				Exporters: []string{"otlp/1"},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	l := plog.NewLogs()
+
+	rl := l.ResourceLogs().AppendEmpty()
+	rl.Resource().Attributes().PutInt("value", 1)
+	rl.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+
+	require.NoError(t, exp.Start(context.Background(), host))
+
+	// test
+	// before #26464, this would panic
+	require.NoError(t, exp.ConsumeLogs(context.Background(), l))
+
+	// verify
+	assert.Len(t, defaultExp.AllLogs(), 1)
+	assert.Len(t, firstExp.AllLogs(), 0)
+}
+
 type mockLogsExporter struct {
 	mockComponent
 	consumertest.LogsSink

--- a/processor/routingprocessor/metrics.go
+++ b/processor/routingprocessor/metrics.go
@@ -106,9 +106,9 @@ func (p *metricsProcessor) route(ctx context.Context, tm pmetric.Metrics) error 
 		rmetrics := tm.ResourceMetrics().At(i)
 		mtx := ottldatapoint.NewTransformContext(
 			nil,
-			pmetric.Metric{},
-			pmetric.MetricSlice{},
-			pcommon.InstrumentationScope{},
+			pmetric.NewMetric(),
+			pmetric.NewMetricSlice(),
+			pcommon.NewInstrumentationScope(),
 			rmetrics.Resource(),
 		)
 

--- a/processor/routingprocessor/metrics_test.go
+++ b/processor/routingprocessor/metrics_test.go
@@ -471,3 +471,46 @@ func TestMetricsAreCorrectlySplitPerResourceAttributeRoutingWithOTTL(t *testing.
 		assert.Equal(t, attr.Double(), float64(-1.0))
 	})
 }
+
+// see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26462
+func TestMetricsAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
+	// prepare
+	defaultExp := &mockMetricsExporter{}
+	firstExp := &mockMetricsExporter{}
+
+	host := newMockHost(map[component.DataType]map[component.ID]component.Component{
+		component.DataTypeMetrics: {
+			component.NewID("otlp"):              defaultExp,
+			component.NewIDWithName("otlp", "1"): firstExp,
+		},
+	})
+
+	exp, err := newMetricProcessor(noopTelemetrySettings, &Config{
+		DefaultExporters: []string{"otlp"},
+		Table: []RoutingTableItem{
+			{
+				Statement: `route() where attributes["value"] > 0`,
+				Exporters: []string{"otlp/1"},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	m := pmetric.NewMetrics()
+
+	rm := m.ResourceMetrics().AppendEmpty()
+	rm.Resource().Attributes().PutInt("value", 1)
+	metric := rm.ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
+	metric.SetEmptyGauge()
+	metric.SetName("cpu")
+
+	require.NoError(t, exp.Start(context.Background(), host))
+
+	// test
+	// before #26464, this would panic
+	require.NoError(t, exp.ConsumeMetrics(context.Background(), m))
+
+	// verify
+	assert.Len(t, defaultExp.AllMetrics(), 1)
+	assert.Len(t, firstExp.AllMetrics(), 0)
+}

--- a/processor/routingprocessor/traces.go
+++ b/processor/routingprocessor/traces.go
@@ -106,8 +106,8 @@ func (p *tracesProcessor) route(ctx context.Context, t ptrace.Traces) error {
 	for i := 0; i < t.ResourceSpans().Len(); i++ {
 		rspans := t.ResourceSpans().At(i)
 		stx := ottlspan.NewTransformContext(
-			ptrace.Span{},
-			pcommon.InstrumentationScope{},
+			ptrace.NewSpan(),
+			pcommon.NewInstrumentationScope(),
 			rspans.Resource(),
 		)
 

--- a/processor/routingprocessor/traces_test.go
+++ b/processor/routingprocessor/traces_test.go
@@ -457,7 +457,9 @@ func TestTracesAreCorrectlySplitPerResourceAttributeWithOTTL(t *testing.T) {
 	})
 }
 
+// see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26462
 func TestAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
+	// prepare
 	defaultExp := &mockTracesExporter{}
 	firstExp := &mockTracesExporter{}
 
@@ -486,8 +488,12 @@ func TestAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
 	span.SetName("span")
 
 	require.NoError(t, exp.Start(context.Background(), host))
+
+	// test
+	// before #26464, this would panic
 	require.NoError(t, exp.ConsumeTraces(context.Background(), tr))
 
+	// verify
 	assert.Len(t, defaultExp.AllTraces(), 1)
 	assert.Len(t, firstExp.AllTraces(), 0)
 

--- a/processor/routingprocessor/traces_test.go
+++ b/processor/routingprocessor/traces_test.go
@@ -457,6 +457,42 @@ func TestTracesAreCorrectlySplitPerResourceAttributeWithOTTL(t *testing.T) {
 	})
 }
 
+func TestAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
+	defaultExp := &mockTracesExporter{}
+	firstExp := &mockTracesExporter{}
+
+	host := newMockHost(map[component.DataType]map[component.ID]component.Component{
+		component.DataTypeTraces: {
+			component.NewID("otlp"):              defaultExp,
+			component.NewIDWithName("otlp", "1"): firstExp,
+		},
+	})
+
+	exp, err := newTracesProcessor(noopTelemetrySettings, &Config{
+		DefaultExporters: []string{"otlp"},
+		Table: []RoutingTableItem{
+			{
+				Statement: `route() where attributes["value"] > 0`,
+				Exporters: []string{"otlp/1"},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	tr := ptrace.NewTraces()
+	rl := tr.ResourceSpans().AppendEmpty()
+	rl.Resource().Attributes().PutInt("value", 1)
+	span := rl.ScopeSpans().AppendEmpty().Spans().AppendEmpty()
+	span.SetName("span")
+
+	require.NoError(t, exp.Start(context.Background(), host))
+	require.NoError(t, exp.ConsumeTraces(context.Background(), tr))
+
+	assert.Len(t, defaultExp.AllTraces(), 1)
+	assert.Len(t, firstExp.AllTraces(), 0)
+
+}
+
 func TestTraceProcessorCapabilities(t *testing.T) {
 	// prepare
 	config := &Config{

--- a/processor/routingprocessor/traces_test.go
+++ b/processor/routingprocessor/traces_test.go
@@ -458,7 +458,7 @@ func TestTracesAreCorrectlySplitPerResourceAttributeWithOTTL(t *testing.T) {
 }
 
 // see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26462
-func TestAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
+func TestTracesAttributeWithOTTLDoesNotCauseCrash(t *testing.T) {
 	// prepare
 	defaultExp := &mockTracesExporter{}
 	firstExp := &mockTracesExporter{}

